Load libraries and data used in week 1

In [1]:
import numpy as np
import pandas as pd

In [2]:
pip install -U scikit-learn

Note: you may need to restart the kernel to use updated packages.


In [3]:
path = "C:\\Users\\z3066824\\OneDrive - UNSW\Documents\\"

In [4]:
file = "train.csv"

In [5]:
df = pd.read_csv(path+file)

In [6]:
df

Id  GP   MIN   PTS  FGM  FGA   FG%  3P Made  3PA   3P%  ...  FTA  \
0     10556  80  24.3   7.8  3.0  6.4  45.7      0.1  0.3  22.6  ...  2.9   
1      5342  75  21.8  10.5  4.2  7.9  55.1     -0.3 -1.0  34.9  ...  3.6   
2      5716  85  19.1   4.5  1.9  4.5  42.8      0.4  1.2  34.3  ...  0.6   
3     13790  63  19.1   8.2  3.5  6.7  52.5      0.3  0.8  23.7  ...  1.5   
4      5470  63  17.8   3.7  1.7  3.4  50.8      0.5  1.4  13.7  ...  0.5   
...     ...  ..   ...   ...  ...  ...   ...      ...  ...   ...  ...  ...   
7995   2996  32   9.2   1.8  0.7  1.8  40.3     -0.1 -0.2  23.1  ...  0.6   
7996  11679  54   6.0   1.8  0.7  1.4  48.7      0.1  0.1   3.1  ...  0.4   
7997   5537  85  28.2  10.7  4.0  9.0  45.1      0.2  0.6  23.6  ...  3.9   
7998   1292  39   7.7   2.5  1.0  2.3  40.1     -0.3 -0.5  13.3  ...  0.7   
7999   2900  49  19.2   4.8  1.7  5.1  32.6      0.7  2.4  41.3  ...  1.3   

       FT%  OREB  DREB  REB  AST  STL  BLK  TOV  TARGET_5Yrs  
0     72.1   2.2   2.0  3.8  3.2  1.1  0.2  1.6            1  
1     67.8   3.6   3.7  6.6  0.7  0.5  0.6  1.4            1  
2     75.7   0.6   1.8  2.4  0.8  0.4  0.2  0.6            1  
3     66.9   0.8   2.0  3.0  1.8  0.4  0.1  1.9            1  
4     54.0   2.4   2.7  4.9  0.4  0.4  0.6  0.7            1  
...    ...   ...   ...  ...  ...  ...  ...  ...          ...  
7995  65.7   0.3   1.8  1.9  0.5  0.3  0.2  0.4            1  
7996  70.1   1.0   1.1  2.0  0.1  0.0  0.3  0.3            1  
7997  69.7   1.0   2.1  3.1  3.4  1.2  0.2  1.8            1  
7998  74.3   0.4   0.6  0.9  0.2  0.3  0.3  0.5            1  
7999  70.1   0.1   1.1  1.2  3.5  0.9 -0.3  1.4            1  

[8000 rows x 21 columns]

Load additional libraries

In [7]:
from numpy import mean
from numpy import std
from numpy import absolute
from pandas import read_csv
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import RepeatedKFold
from sklearn.linear_model import ElasticNet

Define the data

In [8]:
data = df.values

In [9]:
data

array([[ 1.0556e+04,  8.0000e+01,  2.4300e+01, ...,  2.0000e-01,
         1.6000e+00,  1.0000e+00],
       [ 5.3420e+03,  7.5000e+01,  2.1800e+01, ...,  6.0000e-01,
         1.4000e+00,  1.0000e+00],
       [ 5.7160e+03,  8.5000e+01,  1.9100e+01, ...,  2.0000e-01,
         6.0000e-01,  1.0000e+00],
       ...,
       [ 5.5370e+03,  8.5000e+01,  2.8200e+01, ...,  2.0000e-01,
         1.8000e+00,  1.0000e+00],
       [ 1.2920e+03,  3.9000e+01,  7.7000e+00, ...,  3.0000e-01,
         5.0000e-01,  1.0000e+00],
       [ 2.9000e+03,  4.9000e+01,  1.9200e+01, ..., -3.0000e-01,
         1.4000e+00,  1.0000e+00]])

In [10]:
X, y = data[:, :-1], data[:, -1]

Define model

In [11]:
model = ElasticNet(alpha=1.0, l1_ratio=0.5)

Evaluate model - Report average MAE across a 10-fold cross-validation

In [12]:
cv = RepeatedKFold(n_splits=10, n_repeats=3, random_state=1)

In [13]:
scores = cross_val_score(model, X, y, scoring='neg_mean_absolute_error', cv=cv, n_jobs=-1)

In [14]:
scores = absolute(scores)

In [15]:
print('Mean MAE: %.3f (%.3f)' % (mean(scores), std(scores)))

Mean MAE: 0.267 (0.008)


The model achieved a MAE of about 0.267

Now Set up X and y as Numpy arrays

In [16]:
y = np.array(df['TARGET_5Yrs'])

In [17]:
y.shape

(8000,)

In [18]:
X = np.array(df.loc[:,])

In [19]:
X.shape

(8000, 21)

Split data in tran and test sets using sklearn train_test_split package

In [20]:
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error

In [21]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.25)

Set the alpha values

In [22]:
a_min = 0.01
a_max = 0.2
a_num = 10
alphas = np.linspace(a_min, a_max, a_num)

Create the model - set alphas and training on X and ya data

In [23]:
from sklearn.linear_model import LinearRegression

In [24]:
lr = LinearRegression()

In [25]:
for a in alphas:
    eNet_df = ElasticNet(alpha = a).fit(X,y)
    score = eNet_df.score(X,y)
    y_pred = eNet_df.predict(X)
    mse = mean_squared_error(y, y_pred)

In [26]:
print('Alphas{0: .4f}, R2:{1: .2f}, MSE:{2: .2f}, RMSE:{3: .2f}'
      .format(a, score, mse, np.sqrt(mse)))

Alphas 0.2000, R2: 0.30, MSE: 0.10, RMSE: 0.31


In [27]:
model.fit(X, y)

ElasticNet()

Create a tuned model on training data using alpha = 0.2000 

In [50]:
eNet_df_tuned = ElasticNet(alpha = 0.2).fit(X_train, y_train)

In [ ]:
Run prediction on the X_test data

In [52]:
y_pred_tuned = eNet_df_tuned.predict(X_test)

Find model score and mse

In [55]:
score = eNet_df_tuned.score(X_test, y_test)
mse = mean_squared_error(y_test, y_pred_tuned)
print('R2: {0: 4f}, MSE: {1: .2f}, RMSE: {2: .2f}'. format(score, mse,np.sqrt(mse)))

R2:  0.427409, MSE:  0.14, RMSE:  0.38


Run a cross-validation check and examine best alpha, intercept, and coefficent values

In [56]:
from sklearn.linear_model import ElasticNetCV

In [57]:
eNet_df_cv = ElasticNetCV(alphas = alphas, cv = 10)
model = eNet_df_cv.fit(X_train, y_train)

In [58]:
print(model.alpha_)

0.11555555555555555


In [59]:
print(model.intercept_)

0.52967738245748


In [60]:
print(model.coef_)

[ 0.          0.          0.          0.24058788  0.00601113 -0.
 -0.          0.          0.         -0.          0.          0.
 -0.          0.          0.          0.          0.          0.
  0.         -0.        ]


Make predictions from the model

In [88]:
from sklearn.linear_model import LogisticRegression

In [95]:
conda install -c conda-forge imbalanced-learn

Solving environment: ...working... done

## Package Plan ##

  environment location: C:\Users\z3066824\Anaconda3

  added / updated specs:
    - imbalanced-learn


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    ca-certificates-2021.10.8  |       h5b45459_0         176 KB  conda-forge
    certifi-2021.10.8          |   py38haa244fe_1         145 KB  conda-forge
    conda-4.11.0               |   py38haa244fe_0        16.9 MB  conda-forge
    imbalanced-learn-0.9.0     |     pyhd8ed1ab_0         115 KB  conda-forge
    markupsafe-2.1.0           |   py38h294d835_0          25 KB  conda-forge
    openssl-1.1.1l             |       h8ffe710_0         5.7 MB  conda-forge
    python_abi-3.8             |           2_cp38           4 KB  conda-forge
    scikit-learn-1.0.2         |   py38hf11a4ad_1         5.1 MB
    ------------------------------------------------------------
                


The environment is inconsistent, please check the package plan carefully
The following packages are causing the inconsistency:

  - defaults/win-64::anaconda==custom=py38_1
  - defaults/noarch::anaconda-project==0.9.1=pyhd3eb1b0_1
  - defaults/win-64::bokeh==2.3.1=py38haa95532_0
  - defaults/win-64::conda-build==3.21.4=py38haa95532_0
  - defaults/noarch::conda-verify==3.4.2=py_1
  - defaults/noarch::dask==2021.4.0=pyhd3eb1b0_0
  - defaults/noarch::flask==1.1.2=pyhd3eb1b0_0
  - defaults/noarch::ipympl==0.7.0=pyhd3eb1b0_0
  - defaults/noarch::ipywidgets==7.6.3=pyhd3eb1b0_1
  - defaults/noarch::jinja2==2.11.3=pyhd3eb1b0_0
  - defaults/win-64::jupyter==1.0.0=py38_7
  - defaults/noarch::jupyterlab==3.0.14=pyhd3eb1b0_1
  - defaults/noarch::jupyterlab_server==2.4.0=pyhd3eb1b0_0
  - defaults/win-64::jupyter_server==1.4.1=py38haa95532_0
  - defaults/noarch::nbclassic==0.2.6=pyhd3eb1b0_0
  - defaults/win-64::nbconvert==6.0.7=py38_0
  - defaults/win-64::notebook==6.3.0=py38haa95532_0
  - default

openssl-1.1.1l       | 5.7 MB    | ########## | 100% 

conda-4.11.0         | 16.9 MB   |            |   0% 
conda-4.11.0         | 16.9 MB   | 1          |   1% 
conda-4.11.0         | 16.9 MB   | 3          |   3% 
conda-4.11.0         | 16.9 MB   | 5          |   5% 
conda-4.11.0         | 16.9 MB   | 7          |   7% 
conda-4.11.0         | 16.9 MB   | 9          |   9% 
conda-4.11.0         | 16.9 MB   | #1         |  11% 
conda-4.11.0         | 16.9 MB   | #3         |  13% 
conda-4.11.0         | 16.9 MB   | #5         |  16% 
conda-4.11.0         | 16.9 MB   | #7         |  18% 
conda-4.11.0         | 16.9 MB   | #9         |  20% 
conda-4.11.0         | 16.9 MB   | ##1        |  22% 
conda-4.11.0         | 16.9 MB   | ##3        |  23% 
conda-4.11.0         | 16.9 MB   | ##5        |  26% 
conda-4.11.0         | 16.9 MB   | ##7        |  28% 
conda-4.11.0         | 16.9 MB   | ##9        |  30% 
conda-4.11.0         | 16.9 MB   | ###1       |  32% 
conda-4.11.0         | 16.9

In [96]:
from imblearn import under_sampling, over_sampling
from imblearn.over_sampling import SMOTE

In [99]:
from sklearn.model_selection import StratifiedKFold

In [97]:
y_pred = model.predict(X_test)

In [98]:
logistic = LogisticRegression()
prediction = logistic.fit(X, y)

In [106]:
from sklearn.svm import SVC
from sklearn.metrics import confusion_matrix

In [102]:
svm_model= SVC()

In [103]:
svm_model.fit (X_train, y_train)

SVC()

In [104]:
svm_pred= svm_model.predict (X_test)

In [107]:
print (confusion_matrix (y_test, svm_pred))

[[12  1]
 [ 2 10]]


In [110]:
from sklearn.model_selection import GridSearchCV

In [111]:
param_grid = { 'C':[0.1, 1, 10, 100],'gamma': [0.1, 1, 10, 100, 1000]}

In [112]:
grid = GridSearchCV (SVC(), param_grid, verbose=3)

In [113]:
grid.fit (X_train,y_train)

Fitting 5 folds for each of 20 candidates, totalling 100 fits
[CV 1/5] END ..................C=0.1, gamma=0.1;, score=0.467 total time=   0.0s
[CV 2/5] END ..................C=0.1, gamma=0.1;, score=0.467 total time=   0.0s
[CV 3/5] END ..................C=0.1, gamma=0.1;, score=0.867 total time=   0.0s
[CV 4/5] END ..................C=0.1, gamma=0.1;, score=0.467 total time=   0.0s
[CV 5/5] END ..................C=0.1, gamma=0.1;, score=0.667 total time=   0.0s
[CV 1/5] END ....................C=0.1, gamma=1;, score=0.467 total time=   0.0s
[CV 2/5] END ....................C=0.1, gamma=1;, score=0.467 total time=   0.0s
[CV 3/5] END ....................C=0.1, gamma=1;, score=0.667 total time=   0.0s
[CV 4/5] END ....................C=0.1, gamma=1;, score=0.600 total time=   0.0s
[CV 5/5] END ....................C=0.1, gamma=1;, score=0.533 total time=   0.0s
[CV 1/5] END ...................C=0.1, gamma=10;, score=0.467 total time=   0.0s
[CV 2/5] END ...................C=0.1, gamma=10

GridSearchCV(estimator=SVC(),
             param_grid={'C': [0.1, 1, 10, 100],
                         'gamma': [0.1, 1, 10, 100, 1000]},
             verbose=3)

In [114]:
grid.best_params_

{'C': 10, 'gamma': 0.1}

In [115]:
grid_predictions = grid.predict (X_test)

In [117]:
print(y_test, grid_predictions)

[1 1 1 0 0 1 1 0 1 1 1 0 0 0 0 1 0 0 1 1 0 0 1 0 0] [1 0 1 0 0 1 1 0 0 0 1 0 0 1 0 1 0 0 1 1 1 0 1 0 0]


Run additional model diagnostis

1- Estimate UAC

In [100]:
oversample = SMOTE()
X, y = oversample.fit_resample(X_train, y_train)
kfold = StratifiedKFold(n_splits=5, shuffle=True, random_state=2)
scores = cross_val_score(LogisticRegression(max_iter=200), X, y, scoring='roc_auc', cv=kfold)
scores.mean()

0.9477678571428572

A UAC of 0.9477 was attaining meaning reliable predications were attained

2- Check accuracy and model fit to training set

In [64]:
score = model.score(X_test, y_test)
mse = mean_squared_error (y_test, y_pred)
print('R2: {0: 4f}, MSE: {1: .2f}, RMSE: {2: .2f}'. format(score, mse,np.sqrt(mse)))

R2:  0.436903, MSE:  0.14, RMSE:  0.37


An R2 of 0.4369 and MSE of 0.14 indicates that the logistic regression model was a relativey good fit for the training data 

Convert predictions to csv

In [132]:
pd.DataFrame(grid_predictions).to_csv("C:\\Users\\z3066824\\OneDrive - UNSW\Documents\\Blamires_Sean.csv")